src/Sarif.Sarifer/SpamBackgroundAnalyzer.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Text.RegularExpressions;
+using System.Threading;
+
+using Microsoft.CodeAnalysis.Sarif.Driver;
+using Microsoft.CodeAnalysis.Sarif.Writers;
+using Microsoft.CodeAnalysis.SarifPatternMatcher;
+
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.Sarifer
+{
+    [Export(typeof(IBackgroundAnalyzer))]
+    internal class SpamBackgroundAnalyzer : BackgroundAnalyzerBase
+    {
+        private readonly IFileSystem fileSystem;
+        private string currentSolutionDirectory;
+        private ISet<Skimmer<AnalyzeContext>> rules;
+
+        /// <inheritdoc/>
+        public override string ToolName => "Spam";
+
+        /// <inheritdoc/>
+        public override string ToolVersion => "0.1.0";
+
+        /// <inheritdoc/>
+        public override string ToolSemanticVersion => "0.1.0";
+
+        public SpamBackgroundAnalyzer()
+        {
+            this.fileSystem = FileSystem.Instance;
+        }
+
+        protected override void AnalyzeCore(Uri uri, string text, string solutionDirectory, SarifLogger sarifLogger, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(solutionDirectory)
+                || (this.currentSolutionDirectory?.Equals(solutionDirectory, StringComparison.OrdinalIgnoreCase) != true))
+            {
+                // clear older rules
+                this.rules?.Clear();
+                this.currentSolutionDirectory = solutionDirectory;
+
+                if (this.currentSolutionDirectory != null)
+                {
+                    this.rules = LoadSearchDefinitionsFiles(this.fileSystem, this.currentSolutionDirectory);
+                }
+            }
+
+            var disabledSkimmers = new HashSet<string>();
+
+            var context = new AnalyzeContext
+            {
+                TargetUri = uri,
+                FileContents = text,
+                Logger = sarifLogger
+            };
+
+            using (context)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                AnalyzeCommand.AnalyzeTargetHelper(context, this.rules, disabledSkimmers);
+            }
+        }
+
+        internal static ISet<Skimmer<AnalyzeContext>> LoadSearchDefinitionsFiles(IFileSystem fileSystem, string solutionDirectory)
+        {
+            string spamDirectory = Path.Combine(solutionDirectory, ".spam");
+            if (!fileSystem.DirectoryExists(spamDirectory))
+            {
+                return new HashSet<Skimmer<AnalyzeContext>>();
```

Comment:
>HashSet [](start = 27, length = 7)

The reason for the `currentRules` logic was that when you analyze a solution or a project, `AnalyzeCore` will be called once per source file, and we didn't want to reread the rules for each file.

Reply:
That would have been worth a comment.

---
In reply to: [534509623](https://github.com/microsoft/sarif-visualstudio-extension/pull/301#discussion_r534509623) [](ancestors = 534509623)

Reply:
if you check, this method is called in `AnalyzeCore`, so the rules will be created once unless we change the solutionDirectory.

---
In reply to: [534509781](https://github.com/microsoft/sarif-visualstudio-extension/pull/301#discussion_r534509781) [](ancestors = 534509781,534509623)

Reply:
As you pointed out off line: The logic in `AnalyzeCore` prevents this from happening on every file:
```
if (string.IsNullOrEmpty(solutionDirectory)
                || (this.currentSolutionDirectory?.Equals(solutionDirectory, StringComparison.OrdinalIgnoreCase) != true))
```


---
In reply to: [534518965](https://github.com/microsoft/sarif-visualstudio-extension/pull/301#discussion_r534518965) [](ancestors = 534518965,534509781,534509623)